model/validation.go

```diff
@@ -204,16 +204,367 @@ func (t Tags) Validate() error {
 	return errs
 }
 
+// Validate ensures tags are well-formed, but does not insist that tags exist
+func Validate(t map[string]string) error {
+	var errs error
+	for k, v := range t {
+		if !tagRegexp.MatchString(k) {
+			errs = multierror.Append(errs, fmt.Errorf("Invalid tag key: %q", k))
+		}
+		if !tagRegexp.MatchString(v) {
+			errs = multierror.Append(errs, fmt.Errorf("Invalid tag value: %q", v))
+		}
+	}
+	return errs
+}
+
+func validateFQDN(fqdn string) error {
+	if len(fqdn) > 255 {
+		return fmt.Errorf("domain name %q too long (max 255)", fqdn)
+	}
+	if len(fqdn) == 0 {
+		return fmt.Errorf("empty domain name not allowed")
+	}
+
+	for _, label := range strings.Split(fqdn, ".") {
+		if !IsDNS1123Label(label) {
+			return fmt.Errorf("domain name %q invalid (label %q invalid)", fqdn, label)
+		}
+	}
+
+	return nil
+}
+
+// ValidateMatchCondition validates a Match Condition
+func ValidateMatchCondition(mc *proxyconfig.MatchCondition) error {
+	var retVal error
+
+	if mc.Source != "" {
+		if err := validateFQDN(mc.Source); err != nil {
+			retVal = multierror.Append(retVal, err)
+		}
+	}
+
+	if err := Validate(mc.SourceTags); err != nil {
+		retVal = multierror.Append(retVal, err)
+	}
+
+	if mc.GetTcp() != nil {
+		if err := ValidateL4MatchAttributes(mc.GetTcp()); err != nil {
+			retVal = multierror.Append(retVal, err)
+		}
+	}
+
+	if mc.GetUdp() != nil {
+		if err := ValidateL4MatchAttributes(mc.GetUdp()); err != nil {
+			retVal = multierror.Append(retVal, err)
+		}
+	}
+
+	// We do not (yet) validate http_headers.
```

Comment:
We can have either TCP or UDP or HTTP , but not all three. This is another instance of semantics. I think the idea behind the validation code was to validate the semantics of proto and not just the syntax. Http is our main target compared to TCP. Its relatively simple to just check for path/prefix and header being an array.